File: src/IL2C.Runtime/src/System/Runtime/InteropServices/NativePointer.c
////////////////////////////////////////////////////////////////////////////
//
// IL2C - A translator for ECMA-335 CIL/MSIL to C language.
// Copyright (c) Kouji Matsui (@kozy_kekyo, @kekyo@mastodon.cloud)
//
// Licensed under Apache-v2: https://opensource.org/licenses/Apache-2.0
//
////////////////////////////////////////////////////////////////////////////

#include "il2c_private.h"
#include <System/Runtime/InteropServices/NativePointer.h>

/////////////////////////////////////////////////
// VTable and runtime type info declarations

IL2C_RUNTIME_TYPE_BEGIN(
    System_Runtime_InteropServices_NativePointer,
    "System.Runtime.InteropServices.NativePointer",
    IL2C_TYPE_VALUE,
    sizeof(System_Runtime_InteropServices_NativePointer),
    System_ValueType,
    0, 0)
IL2C_RUNTIME_TYPE_END();

File: src/IL2C.Interop/NativePointer.cs
////////////////////////////////////////////////////////////////////////////
//
// IL2C - A translator for ECMA-335 CIL/MSIL to C language.
// Copyright (c) Kouji Matsui (@kozy_kekyo, @kekyo@mastodon.cloud)
//
// Licensed under Apache-v2: https://opensource.org/licenses/Apache-2.0
//
////////////////////////////////////////////////////////////////////////////

using System;

namespace System.Runtime.InteropServices
{
    public struct NativePointer
    {
        private readonly IntPtr pointer;

        private NativePointer(IntPtr value) => this.pointer = value;

        public static implicit operator NativePointer(IntPtr value) => new NativePointer(value);
        public static implicit operator IntPtr(NativePointer value) => value.pointer;
    }
}
